core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/internal/DefaultHttpAuthOption.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.spi.internal;
+
+import java.util.HashMap;
+import java.util.Map;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.auth.spi.HttpAuthOption;
+import software.amazon.awssdk.http.auth.spi.SignerProperty;
+import software.amazon.awssdk.identity.spi.IdentityProperty;
+import software.amazon.awssdk.utils.ToString;
+
+@SdkInternalApi
+public final class DefaultHttpAuthOption implements HttpAuthOption {
+
+    private final String schemeId;
+    private final Map<IdentityProperty<?>, Object> identityProperties;
+    private final Map<SignerProperty<?>, Object> signerProperties;
+
+    DefaultHttpAuthOption(BuilderImpl builder) {
+        this.schemeId = builder.schemeId;
+        this.identityProperties = new HashMap<>(builder.identityProperties);
+        this.signerProperties = new HashMap<>(builder.signerProperties);
+    }
+
+    @Override
+    public String schemeId() {
+        return schemeId;
+    }
+
+    @Override
+    public <T> T identityProperty(IdentityProperty<T> property) {
+        return (T) identityProperties.get(property);
+    }
+
+    @Override
+    public <T> T signerProperty(SignerProperty<T> property) {
+        return (T) signerProperties.get(property);
+    }
+
+    @Override
+    public void forEachIdentityProperty(IdentityPropertyConsumer consumer) {
+        identityProperties.forEach(consumer);
+    }
+
+    @Override
+    public <T> void forEachSignerProperty(SignerPropertyConsumer consumer) {
+        for (SignerProperty<?> p : signerProperties.keySet()) {
+            SignerProperty<T> property = (SignerProperty<T>) p;
+            consumer.accept(property, this.signerProperty(property));
+        }
+    }
+
+    @Override
+    public String toString() {
+        return ToString.builder("HttpAuthOption")
+                       .add("identityProperties", identityProperties)
+                       .add("signerProperties", signerProperties)
+                       .build();
+    }
+
+
+    public static final class BuilderImpl implements Builder {
+        private final String schemeId;
+        private final Map<IdentityProperty<?>, Object> identityProperties = new HashMap<>();
+        private final Map<SignerProperty<?>, Object> signerProperties = new HashMap<>();
+
+        public BuilderImpl(String schemeId) {
```

Comment:
It's possible, but we usually prefer validation of the builder values in the constructor. 
(there is a utility if you've not seen it) `software.amazon.awssdk.utils.Validate`

Reply:
I actually followed this from https://github.com/aws/aws-sdk-java-v2/blob/0d7b693bb49dbd6ade5683f543ca497f0a850420/core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/HttpSignRequest.java#L39 which is how the design for HttpSignRequest called it out. In this case, if an option has no properties, it will be `HttpAuthOption.builder(schemeId).build();` v/s `HttpAuthoption.builder().schemeId(schemeId).build()`.

On validate, I actually need to validate it is not null, even if specified in the `builder()` method directly.

Reply:
It does exist. I have interpreted our existing cases as when there is a type that is required, we'll often use this pattern. Still favor `HttpAuthoption.builder().schemeId(schemeId).build()`.

Reply:
updated